File: tenant-service/src/main/java/com/teasui/tenant/controller/TenantController.java
package com.teasui.tenant.controller;

import com.teasui.common.dto.ApiResponse;
import com.teasui.tenant.dto.CreateTenantRequest;
import com.teasui.tenant.dto.TenantResponse;
import com.teasui.tenant.service.TenantService;
import jakarta.validation.Valid;
import lombok.RequiredArgsConstructor;
import org.springframework.http.HttpStatus;
import org.springframework.http.ResponseEntity;
import org.springframework.web.bind.annotation.*;

import java.util.List;

@RestController
@RequestMapping("/tenants")
@RequiredArgsConstructor
public class TenantController {

    private final TenantService tenantService;

    @PostMapping
    public ResponseEntity<ApiResponse<TenantResponse>> createTenant(
            @Valid @RequestBody CreateTenantRequest request) {
        TenantResponse response = tenantService.createTenant(request);
        return ResponseEntity.status(HttpStatus.CREATED)
                .body(ApiResponse.success(response, "Tenant created successfully"));
    }

    @GetMapping("/{id}")
    public ResponseEntity<ApiResponse<TenantResponse>> getTenantById(@PathVariable String id) {
        TenantResponse response = tenantService.getTenantById(id);
        return ResponseEntity.ok(ApiResponse.success(response));
    }

    @GetMapping("/slug/{slug}")
    public ResponseEntity<ApiResponse<TenantResponse>> getTenantBySlug(@PathVariable String slug) {
        TenantResponse response = tenantService.getTenantBySlug(slug);
        return ResponseEntity.ok(ApiResponse.success(response));
    }

    @GetMapping
    public ResponseEntity<ApiResponse<List<TenantResponse>>> getAllActiveTenants() {
        List<TenantResponse> tenants = tenantService.getAllActiveTenants();
        return ResponseEntity.ok(ApiResponse.success(tenants));
    }

    @PutMapping("/{id}/suspend")
    public ResponseEntity<ApiResponse<TenantResponse>> suspendTenant(@PathVariable String id) {
        TenantResponse response = tenantService.suspendTenant(id);
        return ResponseEntity.ok(ApiResponse.success(response, "Tenant suspended"));
    }

    @PutMapping("/{id}/activate")
    public ResponseEntity<ApiResponse<TenantResponse>> activateTenant(@PathVariable String id) {
        TenantResponse response = tenantService.activateTenant(id);
        return ResponseEntity.ok(ApiResponse.success(response, "Tenant activated"));
    }
}

File: app/src/main/kotlin/com/workflow/app/api/TenantController.kt
package com.workflow.app.api

import com.workflow.api.TenantId
import com.workflow.tenant.*
import jakarta.validation.Valid
import jakarta.validation.constraints.NotBlank
import org.springframework.http.HttpStatus
import org.springframework.http.ResponseEntity
import org.springframework.web.bind.annotation.*

@RestController
@RequestMapping("/api/v1/tenants")
class TenantController(
    private val tenantRegistry: TenantRegistry
) {

    @PostMapping
    fun registerTenant(
        @Valid @RequestBody request: RegisterTenantRequest
    ): ResponseEntity<TenantConfigResponse> {
        val config = TenantConfiguration(
            tenantId = TenantId.of(request.tenantId),
            name = request.name,
            plan = TenantPlan.valueOf(request.plan),
            enabledPlugins = request.enabledPlugins.toSet(),
            featureFlags = request.featureFlags
        )
        tenantRegistry.register(config)
        return ResponseEntity.status(HttpStatus.CREATED).body(TenantConfigResponse.from(config))
    }

    @GetMapping
    fun listTenants(): ResponseEntity<List<TenantConfigResponse>> {
        val tenants = tenantRegistry.getAll().map { TenantConfigResponse.from(it) }
        return ResponseEntity.ok(tenants)
    }

    @GetMapping("/{tenantId}")
    fun getTenant(@PathVariable tenantId: String): ResponseEntity<TenantConfigResponse> {
        val config = tenantRegistry.findOrNull(TenantId.of(tenantId))
            ?: return ResponseEntity.notFound().build()
        return ResponseEntity.ok(TenantConfigResponse.from(config))
    }

    @DeleteMapping("/{tenantId}")
    fun deregisterTenant(@PathVariable tenantId: String): ResponseEntity<Void> {
        if (!tenantRegistry.exists(TenantId.of(tenantId))) {
            return ResponseEntity.notFound().build()
        }
        tenantRegistry.deregister(TenantId.of(tenantId))
        return ResponseEntity.noContent().build()
    }
}

data class RegisterTenantRequest(
    @field:NotBlank val tenantId: String,
    @field:NotBlank val name: String,
    val plan: String = "STARTER",
    val enabledPlugins: List<String> = emptyList(),
    val featureFlags: Map<String, Boolean> = emptyMap()
)

data class TenantConfigResponse(
    val tenantId: String,
    val name: String,
    val plan: String,
    val enabledPlugins: List<String>
) {
    companion object {
        fun from(config: TenantConfiguration) = TenantConfigResponse(
            tenantId = config.tenantId.value,
            name = config.name,
            plan = config.plan.name,
            enabledPlugins = config.enabledPlugins.toList()
        )
    }
}
